src/ngraph/runtime/gpu/cuda_emitter.cpp

```diff
@@ -676,6 +680,236 @@ size_t runtime::gpu::CUDAEmitter::build_elementwise_n_to_1(const GPURuntimeConte
+        writer.block_begin();
+        {
+            writer << "extern __shared__ int dimensions[];\n";
+            writer << "const int tid = blockDim.x*blockIdx.x + threadIdx.x;\n";
+            writer << "if (tid < nthreads)\n";
+            writer.block_begin();
+            {
+                writer << "int dim_product = tid;\n";
+                writer << "int data_idx = 0;\n";
+                writer << "for (int i = threadIdx.x; i < (rank - 1) * " << nthreads_per_block
+                       << "; i += " << nthreads_per_block << ")\n";
+                writer.block_begin();
+                {
+                    writer << "dimensions[i] = division_by_invariant_multiplication(dim_product, "
+                              "dim_magic[data_idx], "
+                              "dim_shift[data_idx]);\n";
+                    writer << "dim_product -= (dimensions[i] * dim_strides[data_idx]);\n";
+                    writer << "data_idx++;\n";
+                }
+                writer.block_end();
+                writer << "dimensions[threadIdx.x + (rank-1) * " << nthreads_per_block
+                       << "] = dim_product;\n";
+                writer << "data_idx = 0;\n";
+                writer << "bool in_bounds = true;\n";
+                writer << "int source_idx = 0;\n";
+                writer << "for (int i = threadIdx.x; i < rank * " << nthreads_per_block
+                       << "; i += " << nthreads_per_block << ")\n";
+                writer.block_begin();
+                {
+                    writer << "int source_di = division_by_invariant_multiplication(dimensions[i], "
+                              "slice_magic[data_idx], "
+                              "slice_shift[data_idx]);\n";
+                    writer << "bool on_stride = (mod16(dimensions[i], source_di, "
+                              "slice_str[data_idx]) == 0);\n";
+                    // within slice of input tensor and a multiple of the slice stride
+                    writer << "bool in_slice_di = (dimensions[i] >= lower_bounds[data_idx]) && "
+                              "(dimensions[i] < upper_bounds[data_idx]) && on_stride;\n";
+                    writer << "in_bounds = in_bounds && in_slice_di;\n";
+                    // subtract off lower bound to convert to source index
+                    writer << "source_di -= lower_bounds[data_idx];\n";
+                    writer << "source_idx += source_di * src_strides[data_idx];\n";
+                    writer << "data_idx++;\n";
+                }
+                writer.block_end();
+                writer << "out[tid] = in_bounds ? source[source_idx] : in[tid];\n";
+            }
+            writer.block_end();
+        }
+        writer.block_end();
+        compiled_kernel = ctx->compiled_kernel_pool->set(kernel_name, writer.get_code());
+    }
+
+    std::vector<int> params;
+    params.reserve(tensor_shape.size() * 3 - 1 + lower_bounds.size() + upper_bounds.size() +
```

Comment:
I think reserve() has a place but this is not that place. Not needed and adds two lines of code.

Reply:
By using reserve one avoids many memcpy's under the hood of the std::vector. The size of the vector can be determined ahead of time and the needed space allocated with reserve so that when push_back is called we never exceed the std::vector's internal storage, which otherwise forces a malloc and memcpy.